src-java/kilda-persistence-neo4j/src/main/java/org/openkilda/persistence/repositories/impl/Neo4jApplicationRepository.java

```diff
@@ -0,0 +1,105 @@
+
+    public Neo4jApplicationRepository(Neo4jSessionFactory sessionFactory, TransactionManager transactionManager) {
+        super(sessionFactory, transactionManager);
+    }
+
+    @Override
+    protected Class<ApplicationRule> getEntityType() {
+        return ApplicationRule.class;
+    }
+
+    @Override
+    public Optional<ApplicationRule> lookupRuleByMatchAndFlow(SwitchId switchId, String flowId, String srcIp,
+                                                              Integer srcPort, String dstIp, Integer dstPort,
+                                                              String proto, String ethType, Long metadata) {
+        Filters filters = getBaseFilters(switchId, srcIp, srcPort, dstIp, dstPort, proto, ethType, metadata);
+        filters.and(new Filter(FLOW_ID_PROPERTY_NAME, EQUALS, flowId));
+        return processQuery(filters);
+    }
+
+    @Override
+    public Optional<ApplicationRule> lookupRuleByMatchAndCookie(SwitchId switchId, Cookie cookie, String srcIp,
+                                                                Integer srcPort, String dstIp, Integer dstPort,
+                                                                String proto, String ethType, Long metadata) {
+        Filters filters = getBaseFilters(switchId, srcIp, srcPort, dstIp, dstPort, proto, ethType, metadata);
+        filters.and(new Filter(COOKIE_PROPERTY_NAME, EQUALS, cookie));
+        return processQuery(filters);
+    }
+
+    @Override
+    public Collection<ApplicationRule> findBySwitchId(SwitchId switchId) {
+        return loadAll(new Filter(SWITCH_ID_PROPERTY_NAME, EQUALS, switchId));
+    }
+
+    @Override
+    public Collection<ApplicationRule> findByFlowId(String flowId) {
+        return loadAll(new Filter(FLOW_ID_PROPERTY_NAME, EQUALS, flowId));
+    }
+
+    private Filters getBaseFilters(SwitchId switchId, String srcIp, Integer srcPort, String dstIp,
+                                               Integer dstPort, String proto, String ethType, Long metadata) {
+        Filters filters = new Filters(new Filter(SWITCH_ID_PROPERTY_NAME, EQUALS, switchId));
+        filters.and(new Filter(SRC_IP_PROPERTY_NAME, EQUALS, srcIp));
+        filters.and(new Filter(SRC_PORT_PROPERTY_NAME, EQUALS, srcPort));
+        filters.and(new Filter(DST_IP_PROPERTY_NAME, EQUALS, dstIp));
+        filters.and(new Filter(DST_PORT_PROPERTY_NAME, EQUALS, dstPort));
+        filters.and(new Filter(PROTO_PROPERTY_NAME, EQUALS, proto));
+        filters.and(new Filter(ETH_TYPE_PROPERTY_NAME, EQUALS, ethType));
+        filters.and(new Filter(METADATA_PROPERTY_NAME, EQUALS, metadata));
+
+        return filters;
+    }
+
+    private Optional<ApplicationRule> processQuery(Filters filters) {
+        Collection<ApplicationRule> results = loadAll(filters);
+
+        if (results.size() > 1) {
+            throw new PersistenceException("Found more that 1 Application Rule entity by criteria");
```

Comment:
I think you mast add more information in error message. At least group ids. 
now it's just like "something is wrong with some group"